xbmc/addons/binary-addons/BinaryAddonManager.cpp

```diff
@@ -0,0 +1,240 @@
+/*
+ *      Copyright (C) 2005-2017 Team Kodi
+ *      http://kodi.tv
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Kodi; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "BinaryAddonManager.h"
+#include "BinaryAddonBase.h"
+
+#include "addons/AddonManager.h"
+#include "threads/SingleLock.h"
+#include "utils/log.h"
+
+using namespace ADDON;
+
+CBinaryAddonManager::~CBinaryAddonManager()
+{
+  CAddonMgr::GetInstance().Events().Unsubscribe(this);
+}
+
+bool CBinaryAddonManager::Init()
+{
+  CAddonMgr::GetInstance().Events().Subscribe(this, &CBinaryAddonManager::OnEvent);
+
+  std::vector<std::tuple<std::string, bool, std::string>> binaryAddonList;
+  if (!CAddonMgr::GetInstance().GetInstalledBinaryAddonPaths(binaryAddonList))
+  {
+    CLog::Log(LOGNOTICE, "Binary Addons: No binary addons present and related manager, init not necessary");
+    return true;
+  }
+
+  for (auto addon : binaryAddonList)
+  {
+    BinaryAddonBasePtr base = std::make_shared<CBinaryAddonBase>();
+    if (!base->Create(std::get<0>(addon)))
+    {
+      CLog::Log(LOGERROR, "Binary Addons: Failed to create base for '%s' and addon not usable", std::get<0>(addon).c_str());
+      continue;
+    }
+    m_installedAddons[base->ID()] = base;
+    if (std::get<1>(addon))
+      m_enabledAddons[base->ID()] = base;
```

Comment:
what's the advantage of having two vectors here, one for all usable installed addons, another for the usable installed enabled addons? The addons contained in the second vector are 100% subset of the addons in the first vector. would it be more efficient to have just one vector containing the addon along with its 'enabled' flag? => header change: <code>BinaryAddonList m_usableAddons;</code>?

Or is there a real performance gain, because the functions using the two vectors are called really very often?